### PR TITLE
Staggered mesh method for k-point restricted periodic MP2 energy calculation

### PR DESCRIPTION
**Pull request by xinxing02**
This PR implements the staggered mesh method for periodic MP2 energy calculation under the spin-restricted setting in our recent paper: _Staggered Mesh Method for Correlation Energy Calculations of Solids: Second-Order Møller–Plesset Perturbation Theory_, J. Chem. Theory Comput. 2021, 17, 8, 4733-4745. 

This method is mainly proposed to reduce the finite size error in MP2 energy calculations. The implementation closely follows the original kmp2 code in pyscf/pyscf/pbc/mp/kmp2.py, and works for both FFTDF and GDF setting.

### CODE REVIEW

File: pyscf/pbc/mp/kmp2_stagger.py [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+
+        #   Construction of orbitals and orbital energies on two staggered meshes
+        nks = get_monkhorst_pack_size(mf.cell, mf.kpts)  #   mesh size
+        self.flag_submesh = flag_submesh
+        if flag_submesh is True:
+            #   Choice 1: Use two staggered submeshes of mf.kpts of half size along each dimension
+            if np.any( nks % 2) :
+                print(
+                    'The k-point mesh in the input mf has odd size in certain dimension.',
+                    'Cannot use its submeshes for staggered mesh calculation!')
```
Comment (fishjojo) [inline pyscf/pbc/mp/kmp2_stagger.py:271]: redundant, please delete

File: pyscf/pbc/mp/kmp2_stagger.py [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+                # the obtained mo_energy and  mo_coeff are not in good agreement with those computed
+                # from an SCF-HF calculation with mesh 1*1*2k.
+                raise NotImplementedError
+
+            half_shift = mf.cell.get_abs_kpts( [0.5/n for n in nks] )
+            kpts_vir = mf.kpts
+            kpts_occ = kpts_vir + half_shift
+            kpts = np.concatenate( (kpts_occ, kpts_vir), axis=0)
+            if isinstance(mf, scf.KRHF):
+                with lib.temporary_env(mf, exxdiv='vcut_sph', with_df=df.FFTDF(mf.cell, mf.kpts)):
```
Comment (fishjojo) [inline pyscf/pbc/mp/kmp2_stagger.py:303]: should exxdiv be consistent with that in the mean field calculation?

Comment (fishjojo) [inline pyscf/pbc/mp/kmp2_stagger.py:303]: Is using FFTDF a must here?

Comment (xinxing02) [reply pyscf/pbc/mp/kmp2_stagger.py:303]: For the use of "exxdiv='vcut_sph', with_df=df.FFTDF", I'm following the comment about band structure calculation in at https://pyscf.org/user/pbc/scf.html?highlight=get_band

**Review by fishjojo — APPROVED** [review]

**Pull request merged**